### PR TITLE
JavaScript code generation for traits

### PR DESCRIPTION
**Pull request by chengluyu**
This PR is WIP. I'll write a summary when it's ready.

### CODE REVIEW

File: shared/src/main/scala/mlscript/JSBackend.scala [hunk 1]
```diff
@@ -223,6 +226,82 @@ class JSBackend {
...
+          )
+      ) :: Nil
+    )
+    const(
+      traitSymbol.lexicalName,
+      JSFuncExpr(
+        N,
+        Nil,
+        Ls(
+          const("tag", id("Object")()),
```
Comment (chengluyu) [inline shared/src/main/scala/mlscript/JSBackend.scala]:
```suggestion
          const("tag", id("Symbol")()),
```

Comment (LPTK) [reply shared/src/main/scala/mlscript/JSBackend.scala]: Didn't you want to use the `Symbol` as a key?

Comment (chengluyu) [reply shared/src/main/scala/mlscript/JSBackend.scala]: Yes! I fixed it!

**Comment by chengluyu** [comment]
We probably need to clean up `JSBackend.translateTraitDeclaration`.

**Review by LPTK — APPROVED** [review]
LGTM!

Do you have a test case that shows matching on a trait constructed on a primitive value?

As in:

```
t1 = T1 1
//│ t1: 1 & t1
//│   = [Number: 1]

case t1 of { T1 -> () }
//│ res: ()
//│    = []
```

You can go ahead and squash-merge.

**Pull request merged**